src/lua/fiber.c

```diff
@@ -667,13 +667,15 @@ static int
 lbox_fiber_cancel(struct lua_State *L)
 {
 	struct fiber *f = lbox_checkfiber(L, 1);
-	fiber_cancel(f);
-	/*
-	 * Check if we're ourselves cancelled.
-	 * This also implements cancel for the case when
-	 * f == fiber().
-	 */
-	luaL_testcancel(L);
+	if (!(f->flags & FIBER_IS_SYSTEM)) {
+		fiber_cancel(f);
+		/*
+		 * Check if we're ourselves cancelled.
+		 * This also implements cancel for the case when
+		 * f == fiber().
+		 */
+		luaL_testcancel(L);
+	}
```

Comment:
I still can cancel such fiber via C API. `fiber_cancel()` is public.

Reply:
> I think that a net.box worker fiber, like any other system fiber, should ignore `fiber.cancel` coming from the Lua world.

If I understood @locker right, we want to limit killing of the fiber only from Lua. From C is should work just the same as it used to do before this patch.

Reply:
AFAICS it's impossible to look up a system fiber handle in C (`fiber_find` isn't public) so even though `fiber_cancel` is public, one can't possibly cancel a system fiber with it. So I think it's okay to leave `fiber_cancel` as is. Anyway, in future we can update it to ignore system fibers if we need to and add `fiber_cancel_system` internal function for stopping system fibers, but now it doesn't seem to be worthwhile.

Reply:
You can reach system fiber, if you make a C function, load it via `box.lib`, and then call it in any netbox trigger. In this C function you can make `fiber_cancel(fiber_self())` and that will kill the system fiber. This entire procedure uses only public API and no FFI hacks.

Reply:
I agree here with Vladislav: it seems counter-intuitive to have different behavior of C's and Lua's fiber cancelling function, especially when the flag itself is a property of C's fiber.

Alexander Lyapunov recently speak on Arch meeting about C and Lua fiber differences and the general point was that it is not nice to have them. I would not increase the gap without a serious reason.

So I ask for the reason of implementing it this way (only for Lua's fiber cancel). Whether it is possible to reach system fibers from C or not, whether it will be possible in a future or not -- those are not reasons to don't implement it in C, those are reason why it is possible to implement it in Lua. IOW, what is the problem with moving the logic to C's `fiber_cancel()`?

Reply:
At least, the `FIBER_IS_SYSTEM` name suggests that it has no relation to C or Lua. If it would be `FIBER_IS_CANCELLABLE_FROM_LUA`, it would not raise those questions.

However, it would raise the question what is the difference between C and Lua here? Both are legal languages to write a module, a storage procedure or an application logic.

Reply:
This PR is actually a step forward - it makes it impossible to cancel a system fiber from normal Lua code, without usage of C modules, thus making Tarantool safer. I agree that we'll probably have to fix C API `fiber_cancel` as well sooner or later, but I see no point in laying off this PR because of this. Let's file a separate ticket for the C API issue and consider fixing it later.